src/tools/r2rdump/R2RMethod.cs

```diff
@@ -45,49 +41,200 @@ class R2RMethod
 
         /// <summary>
         /// The relative virtual address to the unwind info
         /// </summary>
         public int UnwindRVA { get; set; }
 
-        public R2RMethod(byte[] image, ref MetadataReader mdReader, ref MethodDefinition methodDef)
+        public RuntimeFunction(int startRva, int endRva, int unwindRva)
+        {
+            EntryPoint = startRva;
+            Size = endRva - startRva;
+            if (endRva == -1)
+                Size = -1;
+            UnwindRVA = unwindRva;
+        }
+    }
+
+    class R2RMethod
+    {
+        private const int _blockSize = 16;
+
+        public string Name { get; }
+        public SignatureType ReturnType { get; }
+        public SignatureType[] ArgTypes { get; }
+        public int Token { get; }
+
+        public RuntimeFunction NativeCode { get; }
+
+        public R2RMethod(byte[] image, RuntimeFunction[] runtimeFunctions, ref MetadataReader mdReader, MethodDefinitionHandle methodDefHandle, int methodDefEntryPointsOffset)
         {
+            var methodDef = mdReader.GetMethodDefinition(methodDefHandle);
+
             BlobReader signatureReader = mdReader.GetBlobReader(methodDef.Signature);
             SignatureHeader header = signatureReader.ReadSignatureHeader();
             Name = mdReader.GetString(methodDef.Name);
-            ArgCount = signatureReader.ReadCompressedInteger();
+            int argCount = signatureReader.ReadCompressedInteger();
             ReturnType = DecodeSignatureType(ref signatureReader);
-            ArgTypes = new SignatureType[ArgCount];
-            for (int i = 0; i < ArgCount; i++)
+            ArgTypes = new SignatureType[argCount];
+            for (int i = 0; i < argCount; i++)
             {
                 ArgTypes[i] = DecodeSignatureType(ref signatureReader);
             }
+
+            Token = MetadataTokens.GetToken(methodDefHandle);
+            int rid = MetadataTokens.GetRowNumber(methodDefHandle);
+
+            int val = 0;
+            int baseOffset = DecodeUnsigned(image, methodDefEntryPointsOffset, ref val);
+            int nElements = (int)(val >> 2);
+            int entryIndexSize = (int)(val & 3);
+            int offset = MethodDefEntryPointsTryGetAt(image, rid - 1, (int)baseOffset, nElements, entryIndexSize);
+            int id = 0;
+            offset = DecodeUnsigned(image, offset, ref id);
+            if ((id & 1)!=0)
+            {
+                if ((id & 2)!= 0)
+                {
+                    val = 0;
+                    DecodeUnsigned(image, offset, ref val);
+                    offset -= val;
+                }
+                id >>= 2;
+            }
+            else
+            {
+                id >>= 1;
+            }
+            NativeCode = runtimeFunctions[id];
         }
 
-        internal void SetRVA(int startRva, int endRva, int unwindRva)
+        private int MethodDefEntryPointsTryGetAt(byte[] image, int index, int baseOffset, int nElements, int entryIndexSize)
```

Comment:
It may be nice to have the `NativeArray` parser (and `NativeHashtable` too) as separate type, so that it can be used in multiple places.

Reply:
+1